tests/internal/test_module.py

```diff
@@ -0,0 +1,200 @@
+            __import__(name)
+        assert name in sys.modules
+
+        hook.assert_called_once_with(sys.modules[name], 42)
+
+
+def test_import_module_hook_for_module_not_yet_imported():
+    name = "tests.test_module"
+    __import__(name)
+    with unloaded_modules([name]), ModuleWatchdog() as modules:
+        hook = mock.Mock()
+        modules.register_module_hook(name, hook, 42)
+
+        hook.assert_not_called()
+        assert name in modules._hook_map
+
+        assert name not in sys.modules
+        with no_pytest_meta_path_finder():
+            __import__(name)
+            __import__(name)
+        assert name in sys.modules
+
+        hook.assert_called_once_with(sys.modules[name], 42)
+
+
+def test_module_deleted():
+    name = "tests.test_module"
+    path = origin(sys.modules[name])
+    with unloaded_modules([name]), ModuleWatchdog() as modules, no_pytest_meta_path_finder():
+        __import__(name)
+
+        hook = mock.Mock()
+        modules.register_origin_hook(path, hook, 42)
+        modules.register_module_hook(name, hook, 42)
+
+        del sys.modules[name]
+
+        assert path not in modules._origin_map
+
+
+def test_module_unregister_origin_hook():
+    with ModuleWatchdog() as modules:
+        hook = mock.Mock()
+        path = origin(sys.modules[__name__])
+        modules.register_origin_hook(path, hook, 42)
+        modules.register_origin_hook(path, hook, 43)
+        modules.register_origin_hook(path, hook, 44)
+
+        modules.unregister_origin_hook(path, 43)
+
+        with pytest.raises(ValueError):
+            modules.unregister_origin_hook(path, 45)
+
+        assert modules._hook_map[path] == [(hook, 42), (hook, 44)]
+
+
+def test_module_unregister_module_hook():
+    with ModuleWatchdog() as modules:
+        hook = mock.Mock()
+        module = __name__
+        modules.register_module_hook(module, hook, 42)
+        modules.register_module_hook(module, hook, 43)
+        modules.register_module_hook(module, hook, 44)
+
+        modules.unregister_module_hook(module, 43)
+
+        with pytest.raises(ValueError):
+            modules.unregister_module_hook(module, 45)
+
+        assert modules._hook_map[module] == [(hook, 42), (hook, 44)]
+
+
+def test_module_watchdog_multiple_install():
+    n = 3
+    for _ in range(n):
+        ModuleWatchdog.install()
+
+    i = 0
+    while ModuleWatchdog.is_installed() and i < n:
+        ModuleWatchdog.uninstall()
+        i += 1
+
+    assert not ModuleWatchdog.is_installed()
+    assert i == n
+
+
+def test_module_watchdog_subclasses():
+    class MyWatchdog(ModuleWatchdog):
+        pass
+
+    ModuleWatchdog.install()
+    MyWatchdog.install()
+
+    ModuleWatchdog.uninstall()
+    assert not ModuleWatchdog.is_installed()
+
+    MyWatchdog.uninstall()
+    assert not MyWatchdog.is_installed()
+
+    assert not isinstance(sys.modules, ModuleWatchdog)
+
+
+def test_module_import_hierarchy(run_python_code_in_subprocess):
```

Comment:
Aside: we have a few tests of this kind. Would it make sense to invest some time to implement a pytest mark to run these tests in a dedicated interpreter process? The benefit would be that the test can be written as actual Python code rather than a literal string. There is [pytest-forked](https://github.com/pytest-dev/pytest-forked), but works only on platforms with `fork`.

Reply:
I am down for a better pattern!